ReactiveCocoa/Swift/Property.swift

```diff
@@ -8,66 +8,437 @@ public protocol PropertyType {
+	/// when the wrapped property has deinitialized.
+	///
+	/// It is strongly discouraged to use `signal` on any transformed property.
 	public var signal: Signal<Value, NoError> {
 		return _signal()
 	}
-	
+
 	/// Initializes a property as a read-only view of the given property.
 	public init<P: PropertyType where P.Value == Value>(_ property: P) {
+		sources = AnyProperty.capture(property)
+		disposable = nil
 		_value = { property.value }
 		_producer = { property.producer }
 		_signal = { property.signal }
 	}
-	
+
 	/// Initializes a property that first takes on `initialValue`, then each value
 	/// sent on a signal created by `producer`.
 	public init(initialValue: Value, producer: SignalProducer<Value, NoError>) {
-		let mutableProperty = MutableProperty(initialValue)
-		mutableProperty <~ producer
-		self.init(mutableProperty)
+		self.init(propertyProducer: producer.prefix(value: initialValue),
+		          capturing: [])
 	}
-	
+
 	/// Initializes a property that first takes on `initialValue`, then each value
 	/// sent on `signal`.
 	public init(initialValue: Value, signal: Signal<Value, NoError>) {
-		let mutableProperty = MutableProperty(initialValue)
-		mutableProperty <~ signal
-		self.init(mutableProperty)
+		self.init(propertyProducer: SignalProducer(signal: signal).prefix(value: initialValue),
+		          capturing: [])
 	}
-}
 
-extension PropertyType {
-	/// Maps the current value and all subsequent values to a new value.
-	public func map<U>(transform: Value -> U) -> AnyProperty<U> {
-		let mappedProducer = SignalProducer<U, NoError> { observer, disposable in
-			disposable += ActionDisposable { self }
-			disposable += self.producer.map(transform).start(observer)
+	/// Initializes a property by applying the unary `SignalProducer` transform on
+	/// `property`. The resulting property captures `property`.
+	private init<P: PropertyType>(_ property: P, @noescape transform: SignalProducer<P.Value, NoError> -> SignalProducer<Value, NoError>) {
+		self.init(propertyProducer: transform(property.producer),
+		          capturing: AnyProperty.capture(property))
+	}
+
+	/// Initializes a property by applying the binary `SignalProducer` transform on
+	/// `property` and `anotherProperty`. The resulting property captures `property`
+	/// and `anotherProperty`.
+	private init<P1: PropertyType, P2: PropertyType>(_ firstProperty: P1, _ secondProperty: P2, @noescape transform: SignalProducer<P1.Value, NoError> -> SignalProducer<P2.Value, NoError> -> SignalProducer<Value, NoError>) {
+		self.init(propertyProducer: transform(firstProperty.producer)(secondProperty.producer),
+		          capturing: AnyProperty.capture(firstProperty) + AnyProperty.capture(secondProperty))
+	}
+
+	/// Initializes a property from a producer that promises to send at least one
+	/// value synchronously in its start handler before sending any subsequent event.
+	/// If the producer fails its promise, a fatal error would be raised.
```

Comment:
Hmm I have an idea to make it statically safe. But it requires a change to `PropertyType`.

#3040